franklin/github/tests.py

```diff
@@ -1,62 +1,248 @@
+        mock_get.return_value = get_repo
+
+        expected = get_mock_data('api_responses', 'projects')
+        response = self.client.get(url, **self.header)
+
+        self.assertEqual(ordered(expected), ordered(response.data))
+
+    @patch('core.helpers.requests.get')
+    @patch('core.helpers.requests.delete')
+    def test_delete_project(self, mock_delete, mock_get):
+        """ Delete a project that is already registered """
+        url = '/v1/projects/45864453'
+        self.assertEqual(reverse('project_details', args=['45864453', ]), url)
+
+        # Repo details from github
+        get_repo = Mock(status_code=200)
+        get_repo.json.return_value = get_mock_data('github', 'get_repo')
+        mock_get.return_value = get_repo
+
+        # attempts to call github to delete webhook and deploy key
+        mock_delete.return_value = Mock(status_code=204)
+
+        response = self.client.delete(url, **self.header)
+        self.assertEqual(204, response.status_code)
+
+    @patch('core.helpers.requests.get')
+    def test_get_project_details(self, mock_get):
+        """ Returns details for a project that is registered """
+        url = '/v1/projects/45864453'
+        self.assertEqual(reverse('project_details', args=['45864453', ]), url)
+
+        # Repo details from github
+        get_repo = Mock(status_code=200)
+        get_repo.json.return_value = get_mock_data('github', 'get_repo')
+        mock_get.return_value = get_repo
+
+        expected = get_mock_data('api_responses', 'new_site')
+        with patch.dict('os.environ', {'BASE_URL': 'example.com'}):
+            response = self.client.get(url, **self.header)
+
+        self.assertEqual(ordered(expected), ordered(response.data))
+
+    @patch('core.helpers.requests.post')
+    @patch('core.helpers.requests.get')
+    def test_deploy_project(self, mock_get, mock_post):
+        """ Attempts to deploy the latest commit for the project """
+        url = '/v1/projects/45864453/builds'
+        self.assertEqual(reverse('project_builds', args=['45864453', ]), url)
+
+        # Repo details from github
+        get_repo = Mock(status_code=200)
+        get_repo.json.return_value = get_mock_data('github', 'get_repo')
+
+        # branch details from github
+        get_branch = Mock(status_code=200)
+        get_branch.json.return_value = get_mock_data('github', 'get_branch')
+
+        mock_get.side_effect = [get_repo, get_branch]
+
+        # Creating webhook and deploy key on github repo
+        mock_post.return_value = Mock(status_code=200)
+
+        expected = get_mock_data('api_responses', 'build')
+        # Build object has an auto-generated creation date. mock that.
+        with patch('django.utils.timezone.now') as mock_now:
+            mock_now.return_value = datetime(2016, 5, 4)
+            response = self.client.post(url, **self.header)
+
+        self.assertEqual(ordered(expected), ordered(response.data))
```

Comment:
This method is a good example of a "complex" end-to-end test that mocks multiple calls to multiple external dependencies. Also mocks django's auto generated datetime for a model save.